statistics/handle/update.go

```diff
@@ -208,6 +208,114 @@ func (h *Handle) NewSessionStatsCollector() *SessionStatsCollector {
+type SessionIndexUsageCollector struct {
+	sync.Mutex
+
+	mapper  indexUsageMap
+	next    *SessionIndexUsageCollector
+	deleted bool
+}
+
+func (m indexUsageMap) updateByID(id string, value *IndexUsageInformation) {
+	item := m[id]
+	item.QueryCount += value.QueryCount
+	item.RowsSelected += value.RowsSelected
+	if item.LastUsedAt < value.LastUsedAt {
+		item.LastUsedAt = value.LastUsedAt
+	}
+	m[id] = item
+}
+
+func (m indexUsageMap) update(tableSchema string, tableName string, indexName string, value *IndexUsageInformation) {
+	id := fmt.Sprintf("%s.%s.%s", tableSchema, tableName, indexName)
+	m.updateByID(id, value)
+}
+
+func (m indexUsageMap) merge(destMap indexUsageMap) {
+	for id, item := range destMap {
+		m.updateByID(id, &item)
+	}
+}
+
+// Update updates the mapper in SessionIndexUsageCollector.
+func (s *SessionIndexUsageCollector) Update(tableSchema string, tableName string, indexName string, value *IndexUsageInformation) {
+	value.LastUsedAt = time.Now().Format(types.TimeFSPFormat)
+	s.Lock()
+	defer s.Unlock()
+	s.mapper.update(tableSchema, tableName, indexName, value)
+}
+
+// Delete will set s.deleted to true which means it can be deleted from linked list.
+func (s *SessionIndexUsageCollector) Delete() {
+	s.Lock()
+	defer s.Unlock()
+	s.deleted = true
+}
+
+// NewSessionIndexUsageCollector will add a new SessionIndexUsageCollector into linked list headed by idxUsageListHead.
+// idxUsageListHead always points to an empty SessionIndexUsageCollector as a sentinel node. So we let idxUsageListHead.next
+// points to new item. It's helpful to sweepIdxUsageList.
+func (h *Handle) NewSessionIndexUsageCollector() *SessionIndexUsageCollector {
+	h.idxUsageListHead.Lock()
+	defer h.idxUsageListHead.Unlock()
+	newCollector := &SessionIndexUsageCollector{
+		mapper: make(indexUsageMap),
+		next:   h.idxUsageListHead.next,
+	}
+	h.idxUsageListHead.next = newCollector
+	return newCollector
+}
+
+// sweepIdxUsageList will loop over the list, merge each session's local index usage information into handle
+// and remove closed session's collector.
+// For convenience, we keep idxUsageListHead always points to sentinel node. So that we don't need to consider corner case.
+func (h *Handle) sweepIdxUsageList() indexUsageMap {
+	prev := h.idxUsageListHead
+	prev.Lock()
+	mapper := make(indexUsageMap)
+	for curr := prev.next; curr != nil; curr = curr.next {
+		curr.Lock()
+		mapper.merge(curr.mapper)
+		if curr.deleted {
+			prev.next = curr.next
+			curr.Unlock()
+		} else {
+			prev.Unlock()
+			curr.mapper = make(indexUsageMap)
+			prev = curr
+		}
+	}
+	prev.Unlock()
+	return mapper
+}
+
+// DumpIndexUsageToKV will dump in-memory index usage information to KV.
+// For performance considerations, we use `insert` instead of `update`.
+func (h *Handle) DumpIndexUsageToKV() error {
+	mapper := h.sweepIdxUsageList()
+	for id, value := range mapper {
+		idInfo := strings.Split(id, ".")
```

Comment:
It's better to check the length of `idInfo` here otherwise `idInfo[2]` may cause panics.
```
If len(idInfo) != 3 {
    // log
    continue
}
```